src/library/scala/collection/Iterator.scala

```diff
@@ -155,124 +155,115 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
-     *  group was consumed.
+    /** Eagerly fetch size elements to buffer.
+     *
+     *  If buffer is dirty and stepping, copy prefix.
+     *  If skipping, skip ahead.
+     *  Fetch remaining elements.
+     *  If unable to deliver size, then pad if desired.
+     *  Returns true if successful in delivering `count` elements,
+     *  or padded segment, or partial segment.
      */
-    private def takeDestructively(size: Int): Seq[B] = {
-      val buf = new ArrayBuffer[B]
-      var i = 0
-      // The order of terms in the following condition is important
-      // here as self.hasNext could be blocking
-      while (i < size && self.hasNext) {
-        buf += self.next()
-        i += 1
-      }
-      buf
-    }
-
-    private def padding(x: Int) = immutable.ArraySeq.untagged.fill(x)(pad.get())
-    private def gap = (step - size) max 0
-
-    private def go(count: Int) = {
-      val prevSize = buffer.size
-      def isFirst = prevSize == 0
-      // If there is padding defined we insert it immediately
-      // so the rest of the code can be oblivious
-      val xs: Seq[B] = {
-        val res = takeDestructively(count)
-        // was: extra checks so we don't calculate length unless there's reason
-        // but since we took the group eagerly, just use the fast length
-        val shortBy = count - res.length
-        if (shortBy > 0 && pad.isDefined) res ++ padding(shortBy) else res
-      }
-      lazy val len = xs.length
-      lazy val incomplete = len < count
-
-      // if 0 elements are requested, or if the number of newly obtained
-      // elements is less than the gap between sequences, we are done.
-      def deliver(howMany: Int) = {
-        (howMany > 0 && (isFirst || len > gap)) && {
-          if (!isFirst)
-            buffer dropInPlace (step min prevSize)
-
-          val available =
-            if (isFirst) len
-            else howMany min (len - gap)
-
-          buffer ++= (xs takeRight available)
-          filled = true
-          true
+    private def fulfill(): Boolean = {
+      var done = false
+      if (buffer.size > 0)
+        if (step < size)
+          buffer.dropInPlace(step)
+        else if (step > size) {
+          var dropping = step - size
+          while (dropping > 0 && self.hasNext) {
+            self.next(): Unit
+            dropping -= 1
+          }
+          done = dropping > 0
+          buffer.clear()
+        }
+        else
+          buffer.clear()
+      var index = buffer.size
+      if (!done) {
+        while (index < size && self.hasNext) {
+          buffer += self.next()
+          index += 1
+        }
+        if (partial && pad != null) {
+          buffer.sizeHint(size)
+          while (index < size) {
+            buffer += pad()
+            index += 1
+          }
+          // TODO by-name pad should be taken as lazy; fix the test
+          //buffer.padToInPlace(size, pad())
+          //index = size
```

Comment:
Wouldn't that change semantics?

Reply:
Yes but in a good way. It can't possibly be a good idea to invoke the fill arg N times, perhaps.

Reply:
I guess it's very unlikely that someone relies on side effects in filling a grouped iterator 🤷 
I'm not confident to make this decision though.

Reply:
Maybe it needs a special comment for "we can't do this until the hypothetical future breaking version". There is a comment in TreeSeqMap to similar effect. I will change it from overly forceful TODO to something that expresses WBNTHIF.